### PR TITLE
Problem: missing tests and inconsistent behaviour for ZAP corner cases

### LINKED ISSUE #2711: ZAP handler communication errors should be handled consistently
The current implementations of the mechanisms are not consistent with regard to ZAP handler communication errors, most importantly
1. there is no ZAP handler at all
1. the ZAP handler has not yet started / connected to the ZAP socket
1. the ZAP handler has disconnected from the ZAP socket **before** the session tries to bind to the ZAP socket
1. the ZAP handler disconnects from the ZAP socket **after** the session has bound to the ZAP socket
1. the ZAP handler does not read the ZAP request
1. the ZAP handler does not send a ZAP response

The first two error conditions cannot be distinguished. The third could in principle be distinguished from the first and second, but this is not currently implemented, and seems rather arbitrary.

The NULL and CURVE mechanisms simply ignore errors 1-3, and accept the connection attempt. The CURVE mechanism claims to support the Stonehouse pattern by this, however this attributes to error condition 1 only, which appears rather insecure.

The PLAIN mechanism rejects any connection attempt in case of error conditions 1-3. 

Condition 4 will likely result in some server-side error. Error conditions 5 and 6 will probably cause a client-side timeout, which might be acceptable.

I suggest that it can be explicitly configured whether a ZAP handler is required by a dedicated socket option. Inconsistent configurations (e.g. a ZAP domain is defined, but ZAP handler is not configured as required) should result in an error.

Error condition 2 can be avoided by the client code (like it is done in testutil_security.hpp), but it is rather easy to get this wrong (it was also wrong in the tests until some days ago), and hard to notice the effect. Error condition 3 is also easily triggered by errors in the ZAP handler code, and may remain unnoticed, and allow unauthenticated clients.

Should these cases use server assertions (and abort the process), or should they simply emit monitor events? They are not libzmq-internal errors, but still severe process-internal errors, which cannot be signalled synchronously to the libzmq-client code.

### PR DESCRIPTION
**Pull request by sigiesec**
Solution: added tests & aligned behaviour

Fixes #2711

### CODE REVIEW

File: tests/testutil_security.hpp [hunk 1]
```diff
@@ -183,6 +202,8 @@ void zap_handler_generic (void *ctx,
         char *version = s_recv (handler);
         if (!version)
             break; //  Terminating - peer's socket closed
+        if (zap_protocol == zap_disconnect)
+            break;
```
Comment (bluca) [inline tests/testutil_security.hpp]: this leaks the version string

Comment (sigiesec) [reply tests/testutil_security.hpp]: Fixed this in 77f76a4
